src/SetupApi/SetupApi+SP_DRVINFO_DETAIL_DATA.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright © .NET Foundation and Contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace PInvoke
+{
+    using System;
+    using System.Runtime.CompilerServices;
+    using System.Runtime.InteropServices;
+    using System.Runtime.InteropServices.ComTypes;
+
+    /// <content>
+    /// Contains the <see cref="SP_DRVINFO_DETAIL_DATA" /> nested type.
+    /// </content>
+    public partial class SetupApi
+    {
+        /// <summary>
+        /// Contains detailed information about a particular driver information structure.
+        /// </summary>
+        /// <seealso href="https://docs.microsoft.com/en-us/windows/win32/api/setupapi/ns-setupapi-sp_drvinfo_detail_data_w"/>
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Unicode, Pack = 4)]
+        public unsafe struct SP_DRVINFO_DETAIL_DATA
+        {
+            /// <summary>
+            /// The size, in bytes, of the <see cref="SP_DRVINFO_DETAIL_DATA"/> structure.
+            /// </summary>
+            public int cbSize;
+
+            /// <summary>
+            /// Date of the INF file for this driver.
+            /// </summary>
+            public Kernel32.FILETIME InfDate;
+
+            /// <summary>
+            /// The offset, in characters, from the beginning of the <see cref="HardwareID"/> buffer where the CompatIDs list begins.
+            /// </summary>
+            /// <remarks>
+            /// This value can also be used to determine whether there is a hardware ID that precedes the CompatIDs list.
+            /// If this value is greater than 1, the first string in the HardwareID buffer is the hardware ID.
+            /// If this value is less than or equal to 1, there is no hardware ID.
+            /// </remarks>
+            public int CompatIDsOffset;
+
+            /// <summary>
+            /// The length, in characters, of the CompatIDs list starting at offset <see cref="CompatIDsOffset"/> from the beginning of the <see cref="HardwareID"/> buffer.
+            /// </summary>
+            /// <remarks>
+            /// If <see cref="CompatIDsLength"/> is nonzero, the CompatIDs list contains one or more <see langword="null"/>-terminated strings with an additional
+            /// <see langword="null"/> character at the end of the list.
+            /// If <see cref="CompatIDsLength"/> is zero, the CompatIDs list is empty. In that case, there is no additional <see langword="null"/> character at the end of the list.
+            /// </remarks>
+            public IntPtr CompatIDsLength;
```

Comment:
The documentation says that this field is a `DWORD`, not a pointer.

Reply:
Looking at this declaration once more, I believe this is because [pointers (i.e. `ULONG_PTR Reserved` in this case) are aligned to 32-bit (or 4 byte) boundaries on x86 and 64-bit (or 8 byte) boundaries on x64](https://docs.microsoft.com/en-us/windows/win32/midl/c-compiler-packing-issues#natural-alignment).

`cbSize`, `CompatIDsOffset` and `CompatIDsLength` are 4 bytes wide; `FILETIME` 8 bytes, so that's a total of 4 + 8 + 4 + 4 = 20 bytes, which is a multiple of 4 but not of 8.

Reply:
OK - so the solution simply is to use `Pack = 0` (or omit the `Pack` value alltogether).

Reply:
For structs that require specific alignment, Win32 headers tend to do one of the two things - 

- Use `__declspec(align(alignment_size))` or `DECLSPEC_ALIGN` macro. `DirectXMath.h` has several types that leverages this, like `XMMATRIX`. 
- Use padding fields. Quick scan of `setupapi.h` shows that `_SP_SELECTDEVICE_PARAMS_A.Reserved` is one such field. 

Outside of these patterns, its rare to see special alignment considerations pop-up that are purely *inferred* contextually but aren't encoded in the struct-definition in any other manner. It can happen (there are bugs, failure to document, implicit assumptions etc.), but its rare. 

Aside: Sometimes you'll see (C++) `sizeof` return a value larger than the sum of individual member `sizeof` values - which can happen due to default alignment rules associated with individual types. For e.g, `{DWORD, CHAR, DWORD}` would likely return a `sizeof` `12` instead of `9 (4 + 1 + 4)` (`Marshal.SizeOf` usually returns `12` correctly in these situations AFAICT).